### PR TITLE
Allow reactions to pass validations if they only have reaction SMILES

### PR DESCRIPTION
**Pull request by skearnes**
* Adds a new option to ValidationOptions: `allow_reaction_smiles_only`
* When this is true (the default), reactions that have a `REACTION_{CX}SMILES` identifier, zero inputs, and zero outcomes will not be required to have inputs or outcomes. The other checks for e.g. provenance still apply.

### CODE REVIEW

File: ord_schema/validations.py [hunk 1]
```diff
@@ -40,6 +40,8 @@ class ValidationOptions:
     validate_ids: bool = False
     # Require ReactionProvenance for Reactions.
     require_provenance: bool = False
+    # Allow reactions with valid reaction SMILES and nothing else.
+    allow_reaction_smiles_only = True
```
Comment (connorcoley) [inline ord_schema/validations.py:44]: For consistency, use typing here as well?

File: ord_schema/validations.py [hunk 2]
```diff
@@ -410,12 +412,17 @@ def validate_reaction(message: reaction_pb2.Reaction,
...
     if options is None:
         options = ValidationOptions()
-    if len(message.inputs) == 0:
-        warnings.warn('Reactions should have at least 1 reaction input',
-                      ValidationError)
-    if len(message.outcomes) == 0:
-        warnings.warn('Reactions should have at least 1 reaction outcome',
-                      ValidationError)
+    if (options.allow_reaction_smiles_only and
+            message_helpers.get_reaction_smiles(message) and
```
Comment (connorcoley) [inline ord_schema/validations.py:416]: I think we should try to parse the reactants, any agents, and product(s) into RDKit and issue ValidationWarnings if it doesn't parse. Like we do for SMILES in check_compound_identifiers?

File: ord_schema/validations_test.py [hunk 3]
```diff
@@ -175,6 +175,19 @@ def test_reaction(self):
                                     'reaction input'):
             self._run_validation(message)
 
+    def test_reaction_smiles(self):
+        message = reaction_pb2.Reaction()
+        message.identifiers.add(value='test', type='REACTION_SMILES')
+        output = self._run_validation(message)
+        self.assertEmpty(output.errors)
+        self.assertEmpty(output.warnings)
```
Comment (connorcoley) [inline ord_schema/validations_test.py:183]: If adding a Reaction SMILES value check, this warning list will not be empty

File: ord_schema/validations.py (hunk 1 above)
Comment (skearnes) [reply ord_schema/validations.py:44]: ah thanks; done

File: ord_schema/validations.py (hunk 2 above)
Comment (skearnes) [reply ord_schema/validations.py:416]: Right now this branch only executes if there are no reactants/agents or products at all. I'm thinking that I should add a check for valid reaction SMILES and leave this as-is? If there are reactants or products they will be checked in the normal way.

File: ord_schema/validations_test.py (hunk 3 above)
Comment (skearnes) [reply ord_schema/validations_test.py:183]: yes, i'll add this check

File: ord_schema/validations.py (hunk 2 above)
Comment (connorcoley) [reply ord_schema/validations.py:416]: Yes, good idea

**Comment by skearnes** [comment]
Thanks; PTAL.

File: ord_schema/validations.py [hunk 4]
```diff
@@ -446,6 +447,22 @@ def validate_reaction(message: reaction_pb2.Reaction,
 
 def validate_reaction_identifier(message: reaction_pb2.ReactionIdentifier):
     check_type_and_details(message)
+    if Chem and message.type in [
```
Comment (connorcoley) [inline ord_schema/validations.py:450]: Under what circumstances will `Chem` be `None`?

File: ord_schema/validations.py [hunk 5]
```diff
@@ -446,6 +447,22 @@ def validate_reaction(message: reaction_pb2.Reaction,
...
+            smiles = message.value.split()[0]
+        else:
+            smiles = message.value
+        try:
+            rxn = rdChemReactions.ReactionFromSmarts(smiles, useSmiles=True)
+            if rxn is None:
+                warnings.warn(
+                    f'RDKit {RDKIT_VERSION} could not validate'
+                    f' REACTION_{{CX}}SMILES identifier {message.value}',
+                    ValidationError)
```
Comment (connorcoley) [inline ord_schema/validations.py:463]: For consistency with `validate_compound`, which uses `message_helpers.check_compound_identifiers`, let's make this a ValidationWarning not a ValidationError

**Review by connorcoley — APPROVED** [review]

File: ord_schema/validations.py (hunk 4 above)
Comment (skearnes) [reply ord_schema/validations.py:450]: sorry; just copied from validate_compound_identifier. We made rdkit a formal dep, so I'll remove that.

File: ord_schema/validations.py (hunk 5 above)
Comment (skearnes) [reply ord_schema/validations.py:463]: done

Comment (skearnes) [reply ord_schema/validations.py:463]: Actually, we use warning for NAME and error for SMILES (in validate_compound_identifier). The check_compound_identifiers check is less strict, but that's more designed for checking consistency than parse-ability...WDYT?

Comment (skearnes) [reply ord_schema/validations.py:463]: (Just to be clear, the current code will issue a ValidationError if the SMILES can't be parsed for a compound.)

Comment (connorcoley) [reply ord_schema/validations.py:461]: Oh, sorry you're right -- I was going off of name and forgot that was only for a consistency check. In that case, let's stick to ValidationError as you have it

**Pull request merged**